samples/mpsl/timeslot/src/main.c

```diff
@@ -0,0 +1,229 @@
+	.params.normal.hfclk = MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE,
+	.params.normal.priority = MPSL_TIMESLOT_PRIORITY_NORMAL,
+	.params.normal.distance_us = TIMESLOT_REQUEST_DISTANCE_US,
+	.params.normal.length_us = TIMESLOT_LENGTH_US
+};
+
+mpsl_timeslot_signal_return_param_t signal_callback_return_param;
+
+/* FIFO and FIFO data type used for printing the signal type from timeslot callback */
+K_FIFO_DEFINE(cb_fifo);
+struct cb_fifo_data_t {
+	void *fifo_reserved; /* 1st word reserved for use by fifo */
+	int cb_signal_type;
+};
+
+static void error(void)
+{
+	printk("ERROR!\n");
+	dk_set_leds_state(DK_ALL_LEDS_MSK, DK_NO_LEDS_MSK);
+
+	while (true) {
+		/* Spin for ever */
+		k_sleep(1000);
+	}
+}
+
+static mpsl_timeslot_signal_return_param_t *mpsl_timeslot_callback(uint32_t signal_type)
+{
+	mpsl_timeslot_signal_return_param_t *p_ret_val = NULL;
+
+	switch (signal_type) {
+	case MPSL_TIMESLOT_SIGNAL_START:
+		/* Toggle led on every timeslot START */
+		dk_set_led(TIMESLOT_STATUS_LED, (++timeslot_counter) % 2);
+
+		if (request_in_cb) {
+			/* Request new timeslot when callback returns */
+			signal_callback_return_param.params.request.p_next = &timeslot_request_normal;
+			signal_callback_return_param.callback_action = MPSL_TIMESLOT_SIGNAL_ACTION_REQUEST;
+		} else {
+			/* No callback action when callback returns, so timeslot will be ended */
+			signal_callback_return_param.params.request.p_next = NULL;
+			signal_callback_return_param.callback_action = MPSL_TIMESLOT_SIGNAL_ACTION_NONE;
+		}
+
+		p_ret_val = &signal_callback_return_param;
+		break;
+	case MPSL_TIMESLOT_SIGNAL_CANCELLED:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_BLOCKED:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_SESSION_CLOSED:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_SESSION_IDLE:
+		break;
+	case MPSL_TIMESLOT_SIGNAL_INVALID_RETURN:
+		break;
+	default:
+		error();
+		break;
+	}
+
+	struct cb_fifo_data_t cb_data = { .cb_signal_type = signal_type };
+
+	/* Allocate memory for cb_data on heap, and put it in the FIFO. */
+	size_t size = sizeof(struct cb_fifo_data_t);
+	char *mem_ptr = k_malloc(size);
+	__ASSERT_NO_MSG(mem_ptr != 0);
+	memcpy(mem_ptr, &cb_data, size);
+	k_fifo_put(&cb_fifo, mem_ptr);
+
+	return p_ret_val;
+}
+
+static void mpsl_timeslot_demo(void)
+{
+	int err;
+	char input_char;
+
+	printk("-----------------------------------------------------\n");
+	printk("Press a key to open session and request timeslots.\n");
+	printk("* Press 'a' for a session where each request makes a new request\n");
+	printk("* Press 'b' for a session with single requests\n");
+	input_char = console_getchar();
+	printk("%c\n", input_char);
+	err = mpsl_timeslot_session_open(mpsl_timeslot_callback);
+	if (err) {
+		error();
+	}
+
+	if (input_char == 'a') {
+		request_in_cb = true;
+	} else if (input_char == 'b') {
+		request_in_cb = false;
+	}
+
+	err = mpsl_timeslot_request(&timeslot_request_earliest);
+	if (err) {
+		error();
+	}
+
+	printk("Press any key to close the session.\n");
```

Comment:
Don't we need to wait for START or other signals? Or this is up to the user?

Reply:
No, according to the documentation, any current timeslot will be finished before the session is closed. And any scheduled timeslots will be canceled.